File: 2024/4.py
import time

from utils import readInput, ALL_DIRECTIONS

diagonal_pairs: tuple[tuple[int, int], ...] = ((1, 3), (7, 5), (1, 7), (3, 5))

def loadInput(filename: str) -> dict[complex, str]:
    lines = readInput(filename)
    xwords: dict[complex, str] = {}
    for y in range(len(lines)):
        for x in range(len(lines[y])):
            xwords[complex(x, y)] = lines[y][x]
    return xwords

def find_word(xwords: dict[complex, str], start: complex, dir: complex, word: str) -> bool:
    return all(xwords.get(start + i*dir, "") == word[i] for i in range(len(word)))

def part1(xwords: dict[complex, str], word:str="XMAS"):
    occurrencies: int = sum([find_word(xwords, start, dir, word)                    
           for start in xwords if xwords[start] == word[0] for dir in ALL_DIRECTIONS.values()])
    print (f"🎄 Part 1: {occurrencies}", end='')

def part2(xwords: dict[complex, str], word: str="MAS"):
    occurrencies: int = sum([find_word(xwords, mid_pos-ALL_DIRECTIONS[dir1], 
                                       ALL_DIRECTIONS[dir1], word) and
                             find_word(xwords, mid_pos-ALL_DIRECTIONS[dir2], 
                                       ALL_DIRECTIONS[dir2], word)
                             for mid_pos in xwords if xwords[mid_pos] == word[1] 
                             for dir1, dir2 in diagonal_pairs])
    print (f"🎄🎅 Part 2: {occurrencies}", end='')

if __name__ == "__main__":
    title = "Day 4: Ceres Search"
    sub = "❄ "*(len(title)//2+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput("input_4.txt")
    
    t0 = time.time()
    part1(inputs)
    print (" - {:.5f}".format(time.time()-t0))
    
    t0 = time.time()
    part2(inputs)
    print (" - {:.5f}".format(time.time()-t0))

File: 2024/4.jl
ALL_DIRECTIONS = Dict(0=>Complex(0, 1), 1=>Complex(1, 1), 2=>Complex(1, 0),
                  3=>Complex(1, -1), 4=>Complex(0, -1), 5=>Complex(-1 ,-1),
                  6=>Complex(-1, 0), 7=>Complex(-1, 1))

diagonal_pairs = ((1, 3), (7, 5), (1, 7), (3, 5))

function loadInput(filename::String)
    lines = readlines(filename)
    xwords = Dict{Complex, Char}()
    for y in eachindex(lines)
        for x in eachindex(lines[y])
            xwords[Complex(x, y)] = lines[y][x]
        end
    end
    return xwords
end

function find_word(xwords, pos, dir, word)
    all(get(xwords, pos + (i-1)*dir, ' ') == word[i] for i in eachindex(word))
end

function part1(xwords, word="XMAS")
    occurrencies = 0
    for pos in keys(xwords)
        if xwords[pos] == word[1]
            for dir in values(ALL_DIRECTIONS)
                occurrencies += Int(find_word(xwords, pos, dir, word))                    
            end
        end
    end
    println("🎄 Part 1: $occurrencies")
end

function part2(xwords, word="MAS")
    occurrencies = 0
    for mid_pos in keys(xwords) 
        if xwords[mid_pos] == word[2] 
            for (dir1, dir2) in diagonal_pairs
                occurrencies += find_word(xwords, mid_pos-ALL_DIRECTIONS[dir1], ALL_DIRECTIONS[dir1], word) && 
                                find_word(xwords, mid_pos-ALL_DIRECTIONS[dir2], ALL_DIRECTIONS[dir2], word)
            end
        end
    end
    println("🎄🎅 Part 2: $occurrencies")
end

function main()
    title = "Day 4: Ceres Search"
    sub = "❄ " ^(length(title) ÷ 2 + 2)

    println()
    println(" $title ")
    println(sub)

    inputs = loadInput("input_4.txt")

    @time part1(inputs)
    @time part2(inputs)
end

main()